kernel/src/paging.c

```diff
@@ -18,17 +19,57 @@ void set_cr3(PageEntry* P4) {
         : "r"(P4));
 }
 
-PageEntry page_zone_stack[PAGETABLE_SIZE] __attribute__((aligned(4096)));
-PageEntry* page_zone_pointer = &page_zone_stack[0];
+// stores pagetables
+PageEntry page_tables[PAGETABLES_SIZE] __attribute__((aligned(4096)));
+PageEntry* page_tables_top = &page_tables[0];
 
-/*
-    TEMPORARY PAGE ALLOCATION
+// hopefully, less than a quarter of the page stack should only be free at once
+#define FREED_TABLE_SIZE (PAGETABLES_SIZE / 512 / 4)
+PageEntry* free_tables_stack[FREED_TABLE_SIZE];
+uint64_t free_tables_count;
 
-    When heap allocation is implemented, we can actively check if pagetable contains similar
-   children. If a pagetable can be remade as a single large page then all children will be freed and
-   should be available for new page tables.
-*/
-PageEntry* alloc_pagetable() { return page_zone_pointer += 512; }
+// leaks occur if too many tables are freed in total, this will count how many are leaked.
```

Comment:
Is this because of your implementation or something unavoidable?

Reply:
unavoidable without growing buffers.